src/operations/index.ts

```diff
@@ -17,33 +18,29 @@ export function processItems<T>(
   actualPageIndex: number | undefined;
   filteredItemsCount: number | undefined;
 } {
-  let result = items;
-  let pagesCount: number | undefined;
-  let actualPageIndex: number | undefined;
-  let filteredItemsCount: number | undefined;
-
-  if (propertyFiltering) {
-    result = propertyFilter(result, propertyFilteringQuery || { tokens: [], operation: 'and' }, propertyFiltering);
-    filteredItemsCount = result.length;
-  }
-
-  if (filtering) {
-    result = filter(result, filteringText, filtering);
-    filteredItemsCount = result.length;
+  const filterPredicate = composeFilters(
+    createPropertyFilterPredicate(propertyFiltering, propertyFilteringQuery),
+    createFilterPredicate(filtering, filteringText)
+  );
+  if (filterPredicate) {
+    items = items.filter(filterPredicate);
   }
+  const filteredItemsCount = filterPredicate ? items.length : undefined;
 
-  if (sorting) {
-    result = sort(result, sortingState);
+  const comparator = createComparator(sorting, sortingState);
+  if (comparator) {
+    items = items.slice().sort(comparator);
   }
 
-  const allPageResult = result;
   if (pagination) {
-    pagesCount = getPagesCount(result, pagination.pageSize);
-    actualPageIndex = normalizePageIndex(currentPageIndex, pagesCount);
-    result = paginate(result, actualPageIndex, pagination.pageSize);
+    const allPageItems = items;
+    const pagesCount = getPagesCount(items, pagination.pageSize);
+    const actualPageIndex = normalizePageIndex(currentPageIndex, pagesCount);
+    items = paginate(items, actualPageIndex, pagination.pageSize);
+    return { items, allPageItems, pagesCount, actualPageIndex, filteredItemsCount };
```

Comment:
Can we avoid a separate return value in this condition.

It is very inconsistent surprise one of 3 features in this code has a custom return value

Reply:
Thanks, updated

Reply:
Not fixed. You moved the same concern inside `paginate` function.

Expected result that the return value is built directly in this function, so you can see all fields at glance.

---

Pro tip: try to reduce amount of changes and keep the structure closer to the original and it will be much easier to review and ship it